[SPARK-5262] [SPARK-5244] [SQL] add coalesce in SQLParser and widen types for parameters of coalesce

adrian-wang opened the pull request:
I'll add test case in #4040

SparkQA commented:
[Test build #25599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25599/consoleFull) for   PR 4057 at commit [`1b81275`](https://github.com/apache/spark/commit/1b81275a4da4f058a8a27c7d567132748b03ae2a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25599/consoleFull) for   PR 4057 at commit [`1b81275`](https://github.com/apache/spark/commit/1b81275a4da4f058a8a27c7d567132748b03ae2a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25599/
Test PASSed.

yhuai commented:
Thank you for working it!

Can you explain why your change can make `Coalesce(null, 1, null)` work? To me, it looks like an issue in constant folding instead of the return type. 

Also, if we want to parse `Coalesce`, I think we really need to correctly set the return type of it (and correctly set `nullable`). Just taking the first expression's type seems not right. Can you take a look at this issue?

marmbrus commented:
Also, I #4040 can you add the test here?

yhuai commented:
@adrian-wang Seems we need to add rule in `org.apache.spark.sql.catalyst.analysis.HiveTypeCoercion.FunctionArgumentConversion` to cast the data types of expressions in Coalesce. Also, seems `ExpressionEvaluationSuite` will be a good place to add more tests for evaluating Coalesce.

adrian-wang commented:
@yhuai I don't think we need to cast data types, since coalesce() only return the first non-null value from its parameters, so use the first parameter's datatype is OK. I was adding tests in ExpressionEvaluationSuite but was blocked by some code-gen tests, and I am working on that.

@marmbrus it seems #4040 is reverted due to some conflicts... should I port all the code into this PR together?

yhuai commented:
Although coalesce only returns the first non-null value, but we need to correctly set the type of coalesce. Let's say we have two columns. The type of one is LongType and the type of another one is IntType, when we do coalesce, the return type of coalesce should be LongType. Right now, there will be an error when we are trying to call coalesce for these two columns. In FunctionArgumentConversion, we should make sure all expressions of coalesce are in the same type and if they are not, we should add cast.

adrian-wang committed:
cast types for coalesce

adrian-wang committed:
add coalesce in parser

SparkQA commented:
[Test build #26016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26016/consoleFull) for   PR 4057 at commit [`e47c03a`](https://github.com/apache/spark/commit/e47c03a112743973bcb082e73b43db6aea2c8d0e).
 * This patch merges cleanly.

adrian-wang commented:
Yes, I moved my work to FunctionArgumentConversion, and since #4040 is reverted due to conflicts, I added the code together here. So I leave Coalesce() untouched, since we would have the same type in Coalesce for sure. I'll change the title accordingly.

SparkQA commented:
[Test build #26016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26016/consoleFull) for   PR 4057 at commit [`e47c03a`](https://github.com/apache/spark/commit/e47c03a112743973bcb082e73b43db6aea2c8d0e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Rating(userId: Int, movieId: Int, rating: Float, timestamp: Long)`
  * `  case class Movie(movieId: Int, title: String, genres: Seq[String])`
  * `  case class Params(`
  * `class ALS extends Estimator[ALSModel] with ALSParams `
  * `  case class RatingBlock(srcIds: Array[Int], dstIds: Array[Int], ratings: Array[Float]) `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26016/
Test FAILed.

adrian-wang committed:
fix rebase conflicts

SparkQA commented:
[Test build #26017 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26017/consoleFull) for   PR 4057 at commit [`c393e18`](https://github.com/apache/spark/commit/c393e183a315f109a4809ac6741540a40a1660bf).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26017 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26017/consoleFull) for   PR 4057 at commit [`c393e18`](https://github.com/apache/spark/commit/c393e183a315f109a4809ac6741540a40a1660bf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26017/
Test PASSed.

adrian-wang commented:
ping @yhuai

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala:
```diff
@@ -503,6 +503,18 @@ trait HiveTypeCoercion {
       case Sum(e @ TimestampType()) => Sum(Cast(e, DoubleType))
       case Average(e @ TimestampType()) => Average(Cast(e, DoubleType))
+
+      case Coalesce(es) if es.map(_.dataType).distinct.size > 1 =>
```
It will be good to add a comment to explain what we are doing at here.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala:
```diff
@@ -503,6 +503,18 @@ trait HiveTypeCoercion {
+
+      case Coalesce(es) if es.map(_.dataType).distinct.size > 1 =>
+        val dt: Option[DataType] = Some(NullType)
+        val rt = es.map(_.dataType).foldLeft(dt)((r, c) => if (r != None) {
```
Seems using pattern matching on `r` will be better?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala:
```diff
@@ -503,6 +503,18 @@ trait HiveTypeCoercion {
+        })
+        rt match {
+          case Some(ddt) => Coalesce(es.map(Cast(_, ddt)))
+          case None => sys.error("Cannot do type coercion over coalesce")
```
Seems it will be good to also let users know the data types of those expressions in the `Coalesce`. Also, maybe `type coercion` is not very user friendly. How about "Could not determine the return type of Coalesce" and then provide the types of expressions?

yhuai commented:
A few things I feel we should add in this PR.
* Parsing `Coalesce` in HiveQL.
* Unit tests to make sure we can correctly determine the return type (if we do not have these tests)
* Unit tests to make sure we can throw the correct exception when we get cases that we do not handle (for example, complex types).

I think we also need to add a rule in `DecimalPrecision` for `Coalesce`. But, I think it is fine to do it in a follow-up PR (Can you create a PR for it?).

adrian-wang commented:
@yhuai Thanks for your comments! I'll modify this PR accordingly.

adrian-wang committed:
address Yin's comments

SparkQA commented:
[Test build #26385 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26385/consoleFull) for   PR 4057 at commit [`4d0111a`](https://github.com/apache/spark/commit/4d0111a6b5e400d13f280a06d3aa5bfc34005c8b).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26385 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26385/consoleFull) for   PR 4057 at commit [`4d0111a`](https://github.com/apache/spark/commit/4d0111a6b5e400d13f280a06d3aa5bfc34005c8b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26385/
Test PASSed.

marmbrus commented:
Thanks!  Merged to master.
